### PR TITLE
Users/jikuma/kubernetes

### PR DESCRIPTION
**Pull request by jikuma**

### CODE REVIEW

File: Tasks/Kubernetes/src/utilities.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+var https   = require('https');
+var fs      = require('fs');
+import * as path from "path";
+import * as tl from "vsts-task-lib/task";
+import * as os from "os";
+import * as util from "util";
+
+export async function download(url: string, downloadPath: string): Promise<void> {
+    var file = fs.createWriteStream(downloadPath);
```
Comment (kmkumaran) [inline Tasks/Kubernetes/src/utilities.ts]: Would download require retries? Also, I see Packer task does a download (https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/PackerBuild/src/utilities.ts). Would it be better to have this in Common and both the task use it? Adding @bishal-pdMSFT

File: Tasks/Kubernetes/src/clusterconnection.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        this.kubectlPath = tl.which("kubectl", false);
+        this.userDir = utils.getNewUserDirPath();
+    }
+
+    private async initialize() {
+        if(!this.kubectlPath || !fs.existsSync(this.kubectlPath))
+        {
+            tl.debug(tl.loc("DownloadingClient"));
+            this.kubectlPath = path.join(this.userDir, "kubectl") + this.getExecutableExtention();
+            var version = await utils.getStableKubectlVersion();
```
Comment (kmkumaran) [inline Tasks/Kubernetes/src/clusterconnection.ts]: Given that user can provide config file with commands, in future there can be in-compatibility with this specific version and the commands. Can you think about how this experience would be, specifically in hosted agents.

File: Tasks/Kubernetes/src/kubernetesregistrysecret.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+function createSecret(connection: ClusterConnection, authenticationToken: AuthenticationToken, secret: string): any {
+    tl.debug(tl.loc('CreatingSecret', secret));
+    var command = connection.createCommand();
+    command.arg("create")
+    command.arg("secret");
+    command.arg("docker-registry");
+    command.arg(secret);
+    command.arg("--docker-server="+ authenticationToken.getLoginServerUrl());
+    command.arg("--docker-username="+ authenticationToken.getUsername());
+    command.arg("--docker-password="+ authenticationToken.getPassword());
```
Comment (kmkumaran) [inline Tasks/Kubernetes/src/kubernetesregistrysecret.ts]: We have ran into issues in past with other tasks dealing with passwords containing special characters. Can you test them upfront.

File: Tasks/Kubernetes/task.json [hunk 4]
```diff
@@ -0,0 +1,207 @@
+{
+    "id": "CBC316A2-586F-4DEF-BE79-488A1F503564",
+    "name": "Kubernetes",
+    "friendlyName": "Deploy to Kubernetes",
+    "description": "Deploy of run a command on kubernetes cluster",
```
Comment (kmkumaran) [inline Tasks/Kubernetes/task.json]: Do we have to inform user that this task could install kubectl in agent box.

File: Tasks/Kubernetes/task.json [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            "type": "filePath",
+            "label": "Configuration File",
+            "defaultValue": "",
+            "required": true,
+            "helpMarkDown": "Path to the Configuration file or directory.",
+            "visibleRule": "useConfigurationFile = true",
+            "groupName": "commands"
+        },
+        {
+            "name": "arguments",
```
Comment (kmkumaran) [inline Tasks/Kubernetes/task.json]: Please check arguments that can take special characters.

File: Tasks/Kubernetes/task.json [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            "groupName": "output"
+        },
+        {
+            "name": "kubectlOutput",
+            "type": "string",
+            "label": "Kubectl output",
+            "required": false,
+            "defaultValue": "",
+            "groupName": "output",
+            "helpMarkDown": "Name of the variable."
```
Comment (kmkumaran) [inline Tasks/Kubernetes/task.json]: Examples in help mark down will be useful.

File: Tasks/Kubernetes/src/kubernetes.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
+"use strict";
```
Comment (kmkumaran) [inline Tasks/Kubernetes/src/kubernetes.ts]: Can you add L0 tests.

File: Tasks/Kubernetes/task.json (hunk 4 above)
Comment (jikuma) [inline Tasks/Kubernetes/task.json]: we dont install we download kubectl client. I will ask Atul to add this in the Notes

File: Tasks/Kubernetes/task.json [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        {
+            "name": "forceUpdate",
+            "type": "boolean",
+            "label": "Force Update secret",
+            "defaultValue": "true",
+            "helpMarkDown": "Recreate secret.",
+            "groupName": "registry"
+        },
+        {
+            "name": "command",
```
Comment (jikuma) [inline Tasks/Kubernetes/task.json]: Yes

File: Tasks/Kubernetes/task.json (hunk 6 above)
Comment (jikuma) [inline Tasks/Kubernetes/task.json]: I am yet to receive the helptext from Atul. Will update all the help text once i get from Atul.

File: Tasks/Kubernetes/src/kubernetes.ts (hunk 7 above)
Comment (jikuma) [inline Tasks/Kubernetes/src/kubernetes.ts]: Planned to add it next.

File: Tasks/Kubernetes/src/utilities.ts (hunk 1 above)
Comment (bishal-pdMSFT) [inline Tasks/Kubernetes/src/utilities.ts]: Certainly we can move it to common - and enhance it with retries.

Comment (jikuma) [inline Tasks/Kubernetes/src/utilities.ts]: I am moving it to common now. Retries we can do later if required.

File: Tasks/Kubernetes/package.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
+  "dependencies": {
+    "del": "^2.2.0",
```
Comment (jikuma) [inline Tasks/Kubernetes/package.json]: Done

File: Tasks/Kubernetes/task.json [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+            "defaultValue": "",
+            "groupName": "registry"
+        },
+        {
+            "name": "secretName",
+            "type": "string",
+            "label": "Secret name",
+            "required": false,
+            "defaultValue": "",
+            "helpMarkDown": "Secret name",
```
Comment (jikuma) [inline Tasks/Kubernetes/task.json]: As Discussed. No change required.

File: Tasks/Kubernetes/task.json (hunk 8 above)
Comment (jikuma) [inline Tasks/Kubernetes/task.json]: As Discussed. No change required.

File: Tasks/Kubernetes/src/clusterconnection.ts (hunk 2 above)
Comment (jikuma) [inline Tasks/Kubernetes/src/clusterconnection.ts]: we can use an agrument --generator to fix the version of the rest call make by kubectl client.

https://kubernetes.io/docs/user-guide/kubectl-conventions/

File: Tasks/Kubernetes/src/clusterconnection.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        return command.exec(options).fail(error => {
+            errlines.forEach(line => tl.error(line));
+            throw error;
+        });
+    }
+
+    // download kubernetes cluster config file from endpoint
+    private downloadKubeconfigFileFromEndpoint(kubernetesEndpoint: string) : void {
+        this.kubeconfigFile = path.join(this.userDir, "config");
+        var authDetails = tl.getEndpointAuthorization(kubernetesEndpoint, false).parameters;
```
Comment (bishal-pdMSFT) [inline Tasks/Kubernetes/src/clusterconnection.ts]: use getEndpointAuthorizationParameter() instead

Comment (jikuma) [inline Tasks/Kubernetes/src/clusterconnection.ts]: Done

File: Tasks/Kubernetes/task.json (hunk 5 above)
Comment (jikuma) [inline Tasks/Kubernetes/task.json]: verifed that tool runner is handling special character like ' ,'@' etc aant the command execution fails with no such command found error.

**Pull request merged**